causalml/inference/tree/causaltree.pyx

```diff
@@ -234,19 +242,72 @@ class CausalTreeRegressor(object):
-                                           max_depth=self.max_depth,
-                                           min_samples_leaf=self.min_samples_leaf,
-                                           random_state=self.random_state)
-
-        # sample_weight is used to pass the treatment flag
-        self.model.fit(X, y, sample_weight=1 + self.eps * w)
+        # input checking replicated from BaseDecisionTree.fit()
+        random_state = check_random_state(self.random_state)
+        X = check_array(X, dtype=DTYPE, accept_sparse="csc")
+        y = check_array(y, ensure_2d=False, dtype=None)
+        if issparse(X):
+            X.sort_indices()
+
+            if X.indices.dtype != np.intc or X.indptr.dtype != np.intc:
+                raise ValueError("No support for np.int64 index based "
+                                 "sparse matrices")
+
+        y = np.atleast_1d(y)
+        if y.ndim == 1:
+            y = np.reshape(y, (-1, 1))
+        if getattr(y, "dtype", None) != DOUBLE or not y.flags.contiguous:
+            y = np.ascontiguousarray(y, dtype=DOUBLE)
+        n_samples, n_features = X.shape
+        n_outputs = y.shape[1]
+
+        if isinstance(self.min_samples_leaf, numbers.Integral):
+            if not 1 <= self.min_samples_leaf:
+                raise ValueError("min_samples_leaf must be at least 1 "
+                                 "or in (0, 0.5], got %s"
+                                 % self.min_samples_leaf)
+            min_samples_leaf = self.min_samples_leaf
+        else:  # float
+            if not 0. < self.min_samples_leaf <= 0.5:
+                raise ValueError("min_samples_leaf must be at least 1 "
+                                 "or in (0, 0.5], got %s"
+                                 % self.min_samples_leaf)
+            min_samples_leaf = int(ceil(self.min_samples_leaf * n_samples))
+
+        self.tree = Tree(
+            n_features = n_features,
+            # line below is taken from DecisionTreeRegressor.fit method source
+            #   which comments that the tree shouldn't need the n_classes parameter
+            #   but it apparently does
+            n_classes = np.array([1] * n_outputs, dtype=np.intp),
+            n_outputs = n_outputs)
+        splitter = BestSplitter(criterion = CausalMSE(1, X.shape[0]),
+            max_features = n_features,
+            min_samples_leaf = self.min_samples_leaf,
+            min_weight_leaf = 0, # from DecisionTreeRegressor default
+            random_state = self.random_state)
+        # hardcoded values below come from defaults values in
+        #   sklearn.tree._classes.DecisionTreeRegressor
+        builder = DepthFirstTreeBuilder(
+            splitter = splitter,
+            min_samples_split = 2,
+            min_samples_leaf = self.min_samples_leaf,
+            min_weight_leaf = 0,
+            max_depth = self.max_depth,
+            min_impurity_decrease = 0,
+            min_impurity_split = float("-inf"))
```

Comment:
Thanks @peterfoley! I ran the [test notebook](https://gist.github.com/jeongyoonlee/422312757f8c01e825a54b7f71de1ae6) and confirmed that it works the same as before the change. 

Could you please remove `big_number` in `node_impurity()` and `children_impurity()` as well?

Reply:
Oops. I ran it with your `master` instead of this branch. With this branch, the notebook raises an error at Cell 6 as follows:
Code:
```python
# Causal Tree
learner_ct = CausalTreeRegressor(min_samples_leaf=500, random_state=42)
learner_ct.fit(X=X_trn, treatment=w_trn, y=y_trn)
cate_ct = learner_ct.predict(X=X_tst).flatten()
```
Error:
```bash
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-d21a54e6e454> in <module>
      1 # Causal Tree
      2 learner_ct = CausalTreeRegressor(min_samples_leaf=500, random_state=42)
----> 3 learner_ct.fit(X=X_trn, treatment=w_trn, y=y_trn)
      4 cate_ct = learner_ct.predict(X=X_tst).flatten()

causalml/inference/tree/causaltree.pyx in causalml.inference.tree.causaltree.CausalTreeRegressor.fit()

sklearn/tree/_splitter.pyx in sklearn.tree._splitter.Splitter.__cinit__()

TypeError: __cinit__() takes exactly 6 positional arguments (5 given)
```

Reply:
The error above was caused with `sklearn==0.21`, and after upgrading it to `sklearn==0.22`, I got a new error:
```bash
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-d21a54e6e454> in <module>
      1 # Causal Tree
      2 learner_ct = CausalTreeRegressor(min_samples_leaf=500, random_state=42)
----> 3 learner_ct.fit(X=X_trn, treatment=w_trn, y=y_trn)
      4 cate_ct = learner_ct.predict(X=X_tst).flatten()

causalml/inference/tree/causaltree.pyx in causalml.inference.tree.causaltree.CausalTreeRegressor.fit()

sklearn/tree/_tree.pyx in sklearn.tree._tree.DepthFirstTreeBuilder.__cinit__()

TypeError: an integer is required
```

Reply:
I added a couple of comments for changes as below. After those changes, it now runs the notebook fine and produces the compatible results.

Could you please add those fixes? Thanks!